### PR TITLE
Fix: external connect crash

### PR DESCRIPTION
**Pull request by karitra**
There is two async. handlers in `external_t::inner_t::connect`, one for connection, another for connection timeout, latter cancel's the socket, which happen to be unique pointer, the former cancels the timer and resets socket to null on error. But according to the `boost::deadline_timer` [documentation](https://www.boost.org/doc/libs/1_46_1/doc/html/boost_asio/reference/basic_deadline_timer/cancel/overload1.html#boost_asio.reference.basic_deadline_timer.cancel.overload1.remarks)

```
Remarks

If the timer has already expired when cancel() is called, then the handlers for asynchronous wait operations will:

  -   have already been invoked; or
  -   have been queued for invocation in the near future.

These handlers can no longer be cancelled, and therefore are passed an error code that indicates the successful completion of the wait operation. 

```

### TODO:
 - [x] description of race.
 - [x] basic testing.

### CODE REVIEW

File: node/CMakeLists.txt [hunk 1]
```diff
@@ -101,7 +101,7 @@ set(BUILD_FLAGS "${BUILD_FLAGS} -Wold-style-cast")
 set(BUILD_FLAGS "${BUILD_FLAGS} -Woverloaded-virtual")
 set(BUILD_FLAGS "${BUILD_FLAGS} -Wswitch")
 set(BUILD_FLAGS "${BUILD_FLAGS} -Wunreachable-code")
-set(BUILD_FLAGS "${BUILD_FLAGS} -pedantic -pedantic-errors")
+set(BUILD_FLAGS "${BUILD_FLAGS} -pedantic")
```
Comment (shaitan) [inline node/CMakeLists.txt]: Why did you remove `-pedantic-errors`?

File: node/src/isolate/external.cpp [hunk 2]
```diff
@@ -241,60 +254,75 @@ struct external_t::inner_t :
...
         COCAINE_LOG_INFO(log, "connecting to external isolation daemon to {}", boost::lexical_cast<std::string>(endpoint));
+
         auto self_shared = shared_from_this();
+        auto sentry = std::make_shared<flag_sentry_t>(connecting);
+        socket = std::make_unique<tcp::socket>(io_context);
 
         connect_timer.expires_from_now(boost::posix_time::milliseconds(connect_timeout_ms));
-        connect_timer.async_wait([=](const std::error_code& ec){
-            if(!ec) {
-                self_shared->socket->cancel();
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: `self_shared` used to be captured, but now it's not. What guarantees that `this` won't be destroyed before this callback call?

File: node/CMakeLists.txt (hunk 1 above)
Comment (karitra) [reply node/CMakeLists.txt]: To make lambda capture assignment possible, it is a warning with pedantic option for current gcc version. And assignment is needed to pass flag_sentry_t to the asyc. completion handlers without polluting the lambda code by extra variable.

File: node/src/isolate/external.cpp [hunk 3]
```diff
@@ -241,60 +254,75 @@ struct external_t::inner_t :
...
+        connect_timer.async_wait([=, sentry_=sentry](const std::error_code& ec) {
+            // Async handler could be called from another thread.
+            if (ec) {
+                return;
+            }
+
+            COCAINE_LOG_INFO(log, "cancelling connection to isolate daemon");
+            // Just close any last created socket.
+            if (socket) {
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: Why can't it be just as simple as:
* extend this check with `if (socket && socket.get() == expected_socked)` where `expected_socket` is address captured
* replace `socket->cancel();` by `socket->close();` and check in `async_connect` callback that `socket` is still open
* keep previous logic behind `connecting` - only `async_connect` should clear it, but store it as `std::atomic_flag`

Comment (karitra) [reply node/src/isolate/external.cpp]: Reasonable schema, even check `if (socket && socket.get == expected_socket)` could be simplified to `if (socket.get() == expected_socket)`, as `expected_socket` can't be null.

But it seems that there is still a race as socket check for expected and socket->close is not atomic operation. In our case it is not critical as all async. callbacks and `inner_t::connect` operations (except first one from `external_t` constructor) are executed in one thread, but code could be changed someday.

And `connecting` sentry in timer callback serves as kind of global lock on any currently active socket operation, making `inner_t::connect` thread safe, so whatever async. operation will complete last will release the lock.

File: node/src/isolate/external.cpp (hunk 2 above)
Comment (karitra) [reply node/src/isolate/external.cpp]: Yep, self_shared could extends inner_t lifetime beyond external_t lefitme, can't find out if it is desirable, but will fix to preserve original behavior.

File: node/src/isolate/external.cpp [hunk 4]
```diff
@@ -241,60 +286,59 @@ struct external_t::inner_t :
...
-            } else {
-                socket.reset(nullptr);
-                COCAINE_LOG_WARNING(log, "could not connect to external isolation daemon - {}", ec.message());
-                on_fail(ec);
+        const auto connect_timeout_ms = args.as_object().at("connect_timeout_ms", 5000u).as_uint();
+        const auto timeout = boost::posix_time::milliseconds(connect_timeout_ms);
+
+        connection->async_connect(
+            std::move(endpoint),
+            std::move(timeout),
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: `std::move()` on const objets shouldn't work, should it?

File: node/src/isolate/external.cpp [hunk 5]
```diff
@@ -241,60 +286,59 @@ struct external_t::inner_t :
...
+
+        if (connecting_.test_and_set()) {
+            COCAINE_LOG_INFO(log, "connection to isolate daemon is already in progress");
             return;
         }
-        connecting = true;
-        socket.reset(new asio::ip::tcp::socket(io_context));
-        if(session) {
+
+        const auto connection = std::make_shared<connection_t>(shared_from_this(), *log, io_context);
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: Could this allocation be moved close to `async_connect()` call?

File: node/src/isolate/external.cpp [hunk 6]
```diff
@@ -206,27 +207,71 @@ struct external_t::inner_t :
...
+
+        connection_t(const std::shared_ptr<inner_t>& parent, logging::logger_t &log, asio::io_service& io):
+            parent_(parent),
+            log_(log),
+            socket_(std::make_unique<tcp::socket>(io)),
+            connect_timer_(io)
+        {}
+
+        ~connection_t() {
+            parent_->connecting_.clear();
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: Whole `parent` isn't necessary here. And why isn't `connecting_` cleared in `async_completion` callback?

File: node/src/isolate/external.cpp [hunk 7]
```diff
@@ -206,27 +207,71 @@ struct external_t::inner_t :
...
+        }
+
+        auto async_connect(tcp::endpoint endpoint, boost::posix_time::milliseconds timeout, const async_completion callback) -> void {
+            COCAINE_LOG_INFO(log_, "connecting to external isolation daemon at {}", boost::lexical_cast<std::string>(endpoint));
+
+            const auto self = shared_from_this();
+
+            connect_timer_.expires_from_now(timeout);
+            connect_timer_.async_wait([=] (const std::error_code& ec) {
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: Let's explicitly capture values here

File: node/src/isolate/external.cpp [hunk 8]
```diff
@@ -206,27 +207,71 @@ struct external_t::inner_t :
...
+
+        using async_completion = std::function<void(const std::error_code&, std::unique_ptr<tcp::socket>)>;
+
+        const std::shared_ptr<inner_t> parent_;
+        logging::logger_t& log_;
+
+        std::unique_ptr<tcp::socket> socket_;
+        asio::deadline_timer connect_timer_;
+
+        connection_t(const std::shared_ptr<inner_t>& parent, logging::logger_t &log, asio::io_service& io):
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: Let's follow the style for constructors:
```cpp
class_t(arg1, arg2, arg3)
: member1(arg1)
, member2(arg2)
, member3(arg3) {
}
```

Also: `logging::logger_t &log` -> `logging::logger_t& log`

File: node/src/isolate/external.cpp [hunk 9]
```diff
@@ -196,8 +196,9 @@ struct external_t::inner_t :
     public std::enable_shared_from_this<inner_t>
 {
     context_t& context;
+
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: What these spaces are needed for?

File: node/src/isolate/external.cpp [hunk 10]
```diff
@@ -196,8 +196,9 @@ struct external_t::inner_t :
...
     asio::io_service& io_context;
-    std::unique_ptr<tcp::socket> socket;
+
     asio::deadline_timer connect_timer;
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: `connect_timer` should gone.

File: node/src/isolate/external.cpp [hunk 11]
```diff
@@ -206,27 +207,71 @@ struct external_t::inner_t :
...
     static constexpr std::chrono::milliseconds min_seal_time {1000ul};
     static constexpr std::chrono::milliseconds max_seal_time {1000000ul};
 
     std::vector<std::shared_ptr<spool_load_t>> spool_queue;
     std::vector<std::shared_ptr<spawn_load_t>> spawn_queue;
 
+    std::atomic_flag connecting_ = ATOMIC_FLAG_INIT;
+
+    // All async methods should capture shared_from_this()
+    struct connection_t : std::enable_shared_from_this<connection_t> {
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: Lets take `struct connection_t` definition out. Currently it inflates `inner_t`

File: node/src/isolate/external.cpp [hunk 12]
```diff
@@ -206,27 +207,71 @@ struct external_t::inner_t :
...
+                COCAINE_LOG_INFO(log_, "timeout connecting to isolate daemon, ec = {}", ec);
+
+                // socket could be moved out in `async_connect` handler, if not yet,
+                // or on any timer error just close it.
+                if (self->socket_) {
+                    self->socket_->close();
+                }
+            });
+
+            socket_->async_connect(endpoint, [=] (const std::error_code& ec) {
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: Let's explicitly capture values here

File: node/src/isolate/external.cpp [hunk 13]
```diff
@@ -206,27 +207,71 @@ struct external_t::inner_t :
...
+            log_(log),
+            socket_(std::make_unique<tcp::socket>(io)),
+            connect_timer_(io)
+        {}
+
+        ~connection_t() {
+            parent_->connecting_.clear();
+        }
+
+        auto async_connect(tcp::endpoint endpoint, boost::posix_time::milliseconds timeout, const async_completion callback) -> void {
```
Comment (shaitan) [inline node/src/isolate/external.cpp]: Why does it copy `callback` here? And why `callback` is copied into lambda?

File: node/src/isolate/external.cpp [hunk 14]
```diff
@@ -192,41 +192,93 @@ struct metrics_load_t {
...
     std::chrono::time_point<std::chrono::system_clock> last_failed_connect_time;
     std::chrono::milliseconds seal_time;
+
     static constexpr std::chrono::milliseconds min_seal_time {1000ul};
     static constexpr std::chrono::milliseconds max_seal_time {1000000ul};
 
     std::vector<std::shared_ptr<spool_load_t>> spool_queue;
     std::vector<std::shared_ptr<spawn_load_t>> spawn_queue;
 
+    std::atomic_flag connecting_ = ATOMIC_FLAG_INIT;
```
Comment (kletkavrubashku) [inline node/src/isolate/external.cpp]: `connecting_` is public member, do not need to use an underline at the end of name

File: node/CMakeLists.txt (hunk 1 above)
Comment (kletkavrubashku) [reply node/CMakeLists.txt]: Looks like you don't use this feature in current version of PR

File: node/src/isolate/external.cpp (hunk 14 above)
Comment (karitra) [reply node/src/isolate/external.cpp]: Yep, should fix.

File: node/CMakeLists.txt (hunk 1 above)
Comment (karitra) [reply node/CMakeLists.txt]: Need to switch '-pedantic-errors' to allow to compile moved lambda captures:
```cpp
socket_->async_connect(std::move(endpoint), [self, callback=std::move(callback)](const std::error_code& ec) {
// ...
});
```

Comment (kletkavrubashku) [reply node/CMakeLists.txt]: Oh, sorry, I missed that.
